src/Microsoft.Azure.WebJobs.Host/Scale/ScaleMonitorService.cs

```diff
@@ -0,0 +1,149 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using Microsoft.Extensions.Hosting;
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using System.Threading;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+using System.Linq;
+using Microsoft.Extensions.Options;
+using Microsoft.Azure.WebJobs.Hosting;
+
+namespace Microsoft.Azure.WebJobs.Host.Scale
+{
+    /// <summary>
+    /// Class to store scale metrics.
+    /// </summary>
+    internal class ScaleMonitorService : IHostedService, IDisposable
+    {
+        private readonly Timer _timer;
+        private readonly IScaleManager _scaleManager;
+        private readonly IScaleMetricsRepository _metricsRepository;
+        private readonly ILogger _logger;
+        private readonly IOptions<ScaleOptions> _scaleOptions;
+        private readonly IPrimaryHostStateProvider _primaryHostStateProvider;
+        private bool _disposed;
+
+        public ScaleMonitorService(IScaleManager scaleManager, IScaleMetricsRepository metricsRepository, IOptions<ScaleOptions> scaleOptions, IPrimaryHostStateProvider primaryHostStateProvider, ILoggerFactory loggerFactory)
+        {
+            _scaleManager = scaleManager;
+            _metricsRepository = metricsRepository;
+            _timer = new Timer(OnTimer, null, Timeout.Infinite, Timeout.Infinite);
+            _logger = loggerFactory.CreateLogger<ScaleMonitorService>();
+            _scaleOptions = scaleOptions;
+            _primaryHostStateProvider = primaryHostStateProvider;
+        }
+
+        public virtual Task StartAsync(CancellationToken cancellationToken)
+        {
+            // start the timer by setting the due time
+            _logger.LogInformation("Scale monitor service is started.");
+            SetTimerInterval((int)_scaleOptions.Value.ScaleMetricsSampleInterval.TotalMilliseconds);
+
+            return Task.CompletedTask;
+        }
+
+        public Task StopAsync(CancellationToken cancellationToken)
+        {
+            // stop the timer if it has been started
+            _timer.Change(Timeout.Infinite, Timeout.Infinite);
+
+            return Task.CompletedTask;
+        }
+
+        protected async virtual void OnTimer(object state)
+        {
+            if (_primaryHostStateProvider.IsPrimary)
+            {
+                await TakeMetricsSamplesAsync();
+            }
+
+            SetTimerInterval((int)_scaleOptions.Value.ScaleMetricsSampleInterval.TotalMilliseconds);
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+        }
+
+        private async Task TakeMetricsSamplesAsync()
+        {
+            try
+            {
+                // TODO: Need to fix this cast to ScaleManager
```

Comment:
[Mathew]: We need to find a way to remove this cast ideally

Reply:
I guess to avoid the cast we can just define a static helper ScalerManager.GetScalersToSample that takes in the required parameters

Reply:
Making the method static would be problematic as it uses to much stuff from the instance:
```
_targetScalersInError
_monitorManager
_targetScalerManager
_configuration
_scaleOptions
```

My vote is to leave the cast t here.